cura/Machines/Models/QualitySettingsModel.py

```diff
@@ -0,0 +1,148 @@
+        self.addRoleName(self.UserValueRole, "user_value")
+        self.addRoleName(self.CategoryRole, "category")
+
+        self._container_registry = ContainerRegistry.getInstance()
+        self._application = Application.getInstance()
+        self._quality_manager = self._application._quality_manager
+
+        self._extruder_position = ""
+        self._quality_item = None
+        self._i18n_catalog = None
+
+        self._quality_manager.qualitiesUpdated.connect(self._update)
+
+        self._update()
+
+    extruderPositionChanged = pyqtSignal()
+    qualityItemChanged = pyqtSignal()
+
+    def setExtruderPosition(self, extruder_position):
+        if extruder_position != self._extruder_position:
+            self._extruder_position = extruder_position
+            self.extruderPositionChanged.emit()
+            self._update()
+
+    @pyqtProperty(str, fset = setExtruderPosition, notify = extruderPositionChanged)
+    def extruderPosition(self):
+        return self._extruder_position
+
+    def setQualityItem(self, quality_item):
+        if quality_item != self._quality_item:
+            self._quality_item = quality_item
+            self.qualityItemChanged.emit()
+            self._update()
+
+    @pyqtProperty("QVariantMap", fset = setQualityItem, notify = qualityItemChanged)
+    def qualityItem(self):
+        return self._quality_item
+
+    def _update(self):
+        if self._quality_item is None:
+            self.setItems([])
+            return
+
+        items = []
+
+        global_container_stack = Application.getInstance().getGlobalContainerStack()
+        definition_container = global_container_stack.definition
+
+        quality_group = self._quality_item["quality_group"]
+        quality_changes_group = self._quality_item["quality_changes_group"]
+
+        if self._extruder_position == "":
+            quality_node = quality_group.node_for_global
+        else:
+            quality_node = quality_group.nodes_for_extruders.get(self._extruder_position)
+        settings_keys = quality_group.getAllKeys()
+        quality_containers = [quality_node.getContainer()]
+
+        if quality_changes_group is not None:
+            if self._extruder_position == "":
+                quality_changes_node = quality_changes_group.node_for_global
+            else:
+                quality_changes_node = quality_changes_group.nodes_for_extruders.get(self._extruder_position)
+            if quality_changes_node is not None:  # it can be None if number of extruders are changed during runtime
+                try:
+                    quality_containers.insert(0, quality_changes_node.getContainer())
+                except:
+                    # FIXME: This is to prevent incomplete update of QualityManager
+                    return
+            settings_keys.update(quality_changes_group.getAllKeys())
+
+        current_category = ""
+        for definition in definition_container.findDefinitions():
```

Comment:
Isn't it faster to iterate over the settings in the quality?

Reply:
Here we reused the old idea. It iterates over all definitions is because it needs to group settings into categories, and this seems to be the only way to do that.

Reply:
Added doc in 59f31c9fe